packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -116,7 +116,17 @@ class _DaemonServer {
   final NotifyingLogger? notifyingLogger;
 
   Future<void> run() async {
-    final ServerSocket serverSocket = await ServerSocket.bind(InternetAddress.loopbackIPv4, port!);
+    ServerSocket? serverSocket;
+    try {
+      serverSocket = await ServerSocket.bind(InternetAddress.loopbackIPv4, port!);
+    } on SocketException {
+      logger.printTrace('Bind on $port failed with IPv4, retrying on IPv6');
+    }
+
+    // If binding on IPv4 failed, try binding on IPv6.
+    // Omit try catch here, let the failure fallthrough.
+    serverSocket ??= await ServerSocket.bind(InternetAddress.loopbackIPv6, port!);
```

Comment:
LGTM. can you add a test for this? I would:

1. Mark `_DaemonServer` visibleForTesting
2. Rather than directly calling `ServerSocket.bind()`, instead add a `@visibleForTesting` `bind` property on the class, that defaults to `ServerSocket.bind()`
3. Add a test to https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/daemon_test.dart injecting a stateful `bind` callback that throw a SocketException if passed the ipv4 loopback, and then respond with a FakeServerSocket if passed the ipv6 loopback.
4. Have the test assert that we try first with the ipv4 loopback, then fall back to ipv6 after the exception.